### PR TITLE
front: add track column to timestops table

### LINKED ISSUE #9526: Add a "Track" column to schedule tables
### Who would benefit from this feature?

End users

### What is this feature about?

Add a column to the tables (input and output) such that:

- [x] It is named "Track" in English and "Voie" in French
- [x] It is the third column in the table, right after the CH code of the operational points (or requested points)
- [x] For each row in a schedule table (input and output), it is filled with the name of the track the train is on when at this operational point
- [x] The width of the column is fixed (and thus, it takes space on the name of the operational points)

### Why is this feature valuable?

This feature matters because it:
- gives more information to the users about their schedule

### Anything else people should know?

The algorithm to get the name of the track can probably be scavenged from the "fiche train" component which is at the bottom of the scenario detail page today, and which displays a column with the data described above.

### PR DESCRIPTION
**Pull request by Synar**
Close #9526

### CODE REVIEW

**Comment by Synar** [comment]
(Update to the tests coming soon ^^)

**Comment by Synar** [comment]
@Yohh je t'ajoute en review juste pour le mini changement de test

**Review by Akctarus — COMMENTED** [review]
Tested, perhaps we could add a little padding-right for long track names to match the padding-left:

Capture d’écran 2024-11-08 à 11 25 24

Apart from that, LGTM!

**Comment by Synar** [comment]
Thanks! I extended the width a bit.

**Comment by Synar** [comment]
2 notes : - conflict with incoming https://github.com/OpenRailAssociation/osrd/pull/9642 to fix, probably by replacing the formatSuggestedOperationalPoints call by a formatSuggestedOperationalPointsWithTrackName call in use output table data
- we should probably cache the tracks information, possibly as a context in the scenario page

File: front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx
Comment (Synar) [inline front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx:1]: The changes in this file are much more minor than they seem at first glance, we just added a `<ScenarioContextProvider infraId={infra.id} dispatch={dispatch}> `wrapper, but git doesn't catch this due to indentation changes.

**Comment by Synar** [comment]
Todo :
 - [x] add tracksectionsLoading state to hook and context

File: front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx
Comment (SharglutDev) [reply front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx:1]: Ty for clarifying this :)

File: front/src/applications/operationalStudies/hooks/useScenarioContext.tsx
Comment (SharglutDev) [inline front/src/applications/operationalStudies/hooks/useScenarioContext.tsx]: I think we don't need to export this ?

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (SharglutDev) [inline front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Since we are in a hook, can't we use `useInfraId` and `useAppDispatch` directly here so we don't have to pass them in param ?

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (SharglutDev) [inline front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Since we don't export this, maybe we could call this endpoint with `useMutation` in `useCachedTrackSections`'s body directly and call the function here ?

This way we don't need the dispatch. We can also provide an `isLoading` so you don't have to use the extra state.

```suggestion
   const [loadInfraObject, { isLoading }] = osrdEditoastApi.endpoints.postInfraByInfraIdObjectsAndObjectType.useMutation();
```

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (SharglutDev) [inline front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Are we sure `trackSections` will be already updated after the call of `loadTracksSectionsByIds` ?

Do we have to pass by a state for `trackSections` value ? Couldn't we just pass the content of `loadTracksSectionsByIds` in the `try` so we can access the result directly ?

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (SharglutDev) [inline front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Won't this be called twice since we `setTrackIds` at the end but we also put `trackIds` in the deps ?

File: front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts
Comment (SharglutDev) [inline front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts]: This one is not an async function

File: front/src/modules/pathfinding/utils.ts
Comment (SharglutDev) [inline front/src/modules/pathfinding/utils.ts]: Nit : for me, it implies that the function returns only the track names. Maybe we could try this (only suggestion)

```suggestion
export const addTrackNamesToSuggestedOperationalPoints = async (
```

File: front/src/modules/timesStops/TimesStopsInput.tsx
Comment (SharglutDev) [inline front/src/modules/timesStops/TimesStopsInput.tsx]: For readability :)

```suggestion
    };
    
    fetchAndFormatRows();
```

File: front/src/modules/timesStops/hooks/useOutputTableData.ts
Comment (SharglutDev) [inline front/src/modules/timesStops/hooks/useOutputTableData.ts]: Can't we keep the `useMemo` here and avoid the state ? Or the component won't rerender if we don't ?

**Review by SharglutDev — COMMENTED** [review]
Thank you for the last changes ! Left some comments, they are totally debatable

File: front/src/applications/operationalStudies/hooks/useScenarioContext.tsx
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useScenarioContext.tsx]: Indeed thanks!

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Agreed for appDispatch, but clara told me we want to get the infra id out of the store and into the context. And while this is just a preliminary step I would still prefer to avoid adding another reliance to the store for the infra id.

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Perhaps usequery since we only fetch? I'm not sure I fully understand the implications of the choice of using dispatch, usemutation or usequery though, so I'll defer to your judgement 

(Nitpick : that loading will encompass less code though, but the sync code should be instantaneous so it probably doesn't matter at all.)

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (SharglutDev) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Ok seems legit :)

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: I don't believe so since it's a usecallback not a useeffect, but it will indeed redefine the function every time.

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (SharglutDev) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Like some other endpoints from the api, we fetch here but in reality, as we need to pass a body to the call, it's technically a post so we can only do a `useMutation` :)

If we want to use it with a query, we need to override the endpoint like we did for other of them in this file : `openapi-editoast-config.ts`

File: front/src/modules/timesStops/hooks/useOutputTableData.ts
Comment (Synar) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts]: Correct me if I'm wrong on this, but since the format operation is now async and useMemo doesn't natively support async operations, I believe we need to use useEffect to handle it. Then because useEffect does not return a value, I believe we need to use the state to track and store the result (in a way that is rerendered) once it completes.

I don't have a ton of experience with hooks yet so I may be overlooking something though

File: front/src/modules/timesStops/TimesStopsInput.tsx
Comment (Synar) [reply front/src/modules/timesStops/TimesStopsInput.tsx]: Sure! ^^

File: front/src/modules/pathfinding/utils.ts
Comment (Synar) [reply front/src/modules/pathfinding/utils.ts]: Sure! By calling it fetch I wanted to emphasize the api call, but the fact the function is async should be enough clues.

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: > Are we sure trackSections will be already updated after the call of loadTracksSectionsByIds ?

I assumed it would, but indeed it may not be.

> Do we have to pass by a state for trackSections value ? Couldn't we just pass the content of loadTracksSectionsByIds in the try so we can access the result directly ?

We can't just pass the value without also reading the state because we only fetch new ids, but you're right this needs to be reworked a bit.

File: front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts]: Got mixed up, thanks!

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Thanks for the explanation! I ended up going with useMutation.

**Comment by Synar** [comment]
2 new fixup, the second one is the change from usestate to useref for the content of the cache, to avoid dependency loops and useless renders

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Since the function was itself used as a dependency of useffects though, I opted to drop the states and go for useref instead. It seems to work properly when testing.

**Comment by Synar** [comment]
Clara asked me to ping all maintainers about adding context to the scenario, so here : @kmer2016 @emersion @SharglutDev @clarani @Math-R . Sorry for the inconvenience, hope I didn't forget anyone ^^

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (SharglutDev) [inline front/src/applications/operationalStudies/hooks/useTrackSections.ts:1]: Feel like it's always better to have the file and hook name synchronized. Could we chose one or the other ? :)

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (SharglutDev) [inline front/src/applications/operationalStudies/hooks/useTrackSections.ts]: Don't we want to do that in the `try` ?

File: front/src/modules/timesStops/hooks/useOutputTableData.ts
Comment (SharglutDev) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts]: Yep you're right

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts:1]: Done ^^

File: front/src/applications/operationalStudies/hooks/useTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useTrackSections.ts]: You're right! Thanks!

**Comment by SharglutDev** [comment]
Lgtm and tested for me. Waiting for at least a 2nd maintainer approval for the context subject :) (@clarani)

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx
Comment (kmer2016) [inline front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx]: Since `trackSectionsLoading` is only used by `TimeStopInput`, and it already accesses the context for `getTrackSectionsByIds`, it would be cleaner and more maintainable to fetch it directly from the context, avoiding unnecessary prop drilling. What do you think"

**Review by kmer2016 — COMMENTED** [review]
Nice job ! Just a small suggestion

**Comment by emersion** [comment]
Overall the context approach sounds good to me :)

File: front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx
Comment (Synar) [reply front/src/applications/operationalStudies/views/ManageTrainSchedule.tsx]: Sure! I wanted to have timesstopsinput mirror timesstopoutput, but yes it probably makes sense to put it directly in timestopsinput here

File: front/src/modules/pathfinding/utils.ts
Comment (clarani) [inline front/src/modules/pathfinding/utils.ts]: You can use undefined by default here and remove it l66
```suggestion
      : undefined;

    return {
      ...op,
      trackName: trackName,
```

File: front/src/modules/pathfinding/utils.ts
Comment (clarani) [inline front/src/modules/pathfinding/utils.ts]: Hum I don't think it's very useful to give the argument `getTrackSectionsById`, it would be better to call it outside `addTrackNamesToSuggestedOperationalPoints` and to give in argument only `operationalPoints` and `trackSections`.

File: front/src/modules/timesStops/hooks/useOutputTableData.ts
Comment (clarani) [inline front/src/modules/timesStops/hooks/useOutputTableData.ts]: Ah hum the type SuggestedOperationalPoint is very messy and I wish to avoid spreading it. Can you make the function addTrackNames more generic so it takes only operationalPoints which contain at least a trackId and returns the same list of operationalPoints with their trackNames ?

It would be smth like this:
```
export const addTrackNamesToOperationalPoints = async <T extends { track: string}>(
  operationalPoints: T[],
  trackSections: InfraObjectWithGeometry[]
): Promise<(T & { trackName?: string})[]> => {...}
```

File: front/src/modules/pathfinding/utils.ts
Comment (clarani) [inline front/src/modules/pathfinding/utils.ts]: Maybe trackSections can be a Map instead of an array ? What do you think ?

File: front/src/applications/operationalStudies/hooks/useCachedTrackSections.ts
Comment (clarani) [inline front/src/applications/operationalStudies/hooks/useCachedTrackSections.ts]: You can use const directly l18
```suggestion
```

File: front/src/applications/operationalStudies/hooks/useCachedTrackSections.ts
Comment (clarani) [inline front/src/applications/operationalStudies/hooks/useCachedTrackSections.ts]: Do you think it would be smart to store only the railjson, so the type would be directly `TrackSection[]` ?

File: front/src/applications/operationalStudies/hooks/useScenarioContext.tsx
Comment (clarani) [inline front/src/applications/operationalStudies/hooks/useScenarioContext.tsx]:
```suggestion
    throw new Error('useScenarioContext must be used within a ScenarioContextProvider');
```

File: front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx
Comment (clarani) [inline front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx:1]: I think you can even move the ScenarioContextProvider in the parent component, in Scenario.tsx

File: front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts
Comment (clarani) [inline front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts:1]: Nice :)

File: front/src/modules/simulationResult/hooks/useFormattedOperationalPoints.ts
Comment (clarani) [inline front/src/modules/simulationResult/hooks/useFormattedOperationalPoints.ts]: Would it be possible to get the `trackSections` outside the function `formatOperationalPoints` and to only give the trackSections as argument ?

File: front/src/modules/pathfinding/utils.ts
Comment (Synar) [reply front/src/modules/pathfinding/utils.ts]: Sure!

File: front/src/modules/pathfinding/utils.ts
Comment (Synar) [reply front/src/modules/pathfinding/utils.ts]: Honestly I just duplicated the old code for this, but sure!

File: front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx
Comment (Synar) [reply front/src/applications/operationalStudies/components/Scenario/ScenarioContent.tsx:1]: Ok!

File: front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useSetupItineraryForTrainUpdate.ts:1]: Thanks!

File: front/src/modules/simulationResult/hooks/useFormattedOperationalPoints.ts
Comment (Synar) [reply front/src/modules/simulationResult/hooks/useFormattedOperationalPoints.ts]: I think so!

File: front/src/applications/operationalStudies/hooks/useScenarioContext.tsx
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useScenarioContext.tsx]: Nice catch thanks!

File: front/src/modules/timesStops/hooks/useOutputTableData.ts
Comment (Synar) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts]: I thought about that, got lazy and decided to avoid messing with generics, but yeah you're right I will

File: front/src/modules/pathfinding/utils.ts
Comment (Synar) [reply front/src/modules/pathfinding/utils.ts]: Nice catch! I don't understand why I wrote that in the first place ^^''

Comment (Synar) [reply front/src/modules/pathfinding/utils.ts]: Actually let me simplify this further
`    const trackName = (
      associatedTrackSection?.railjson as TrackSectionEntity['properties'] | undefined
    )?.extensions?.sncf?.track_name;`

File: front/src/modules/pathfinding/utils.ts
Comment (Synar) [reply front/src/modules/pathfinding/utils.ts]: It does mean that every caller will have to perform the 3 same steps instead of just calling the function though :
-> extract ids to fetch
-> fetch the ids
-> call this function (which is reduced to its last step)

Plus if we need to make it generic, we may as well delete the function and just inline everything

File: front/src/applications/operationalStudies/hooks/useCachedTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useCachedTrackSections.ts]: Done!

File: front/src/applications/operationalStudies/hooks/useCachedTrackSections.ts
Comment (Synar) [reply front/src/applications/operationalStudies/hooks/useCachedTrackSections.ts]: Good catch thanks! I used to use it outside the try, but there's no longer a reason to declare it before.

File: front/src/modules/timesStops/hooks/useOutputTableData.ts
Comment (clarani) [inline front/src/modules/timesStops/hooks/useOutputTableData.ts]: You don't need to do this map ;)
You can directly get the trackName of the op l114 and return it in the correct object
```suggestion
```

File: front/src/modules/pathfinding/utils.ts
Comment (clarani) [reply front/src/modules/pathfinding/utils.ts]: Yes you're right. I find the code way cleaner and easier to understand, like this, thanks :star_struck:

File: front/src/modules/timesStops/hooks/useOutputTableData.ts
Comment (Synar) [reply front/src/modules/timesStops/hooks/useOutputTableData.ts]: Oh right thanks! I was still thinking in terms of the previous function, but there's no longer a need for that ^^

**Review by Maymanaf — COMMENTED** [review]
Nice PR! The changes look good for the minor update in the e2e test.

If possible, could you expand the scope to include a check for the track values? If adding it now isn't feasible, please create an issue and assign it to me for follow-up later.

Thanks! 😊

**Comment by Synar** [comment]
> Nice PR! The changes look good for the minor update in the e2e test.
> 
> If possible, could you expand the scope to include a check for the track values? If adding it now isn't feasible, please create an issue and assign it to me for follow-up later.
> 
> Thanks! 😊

Done!

**Review by clarani — APPROVED** [review]
LGTM :white_check_mark:

**Review by Maymanaf — APPROVED** [review]
LGTM for e2e tests  ✅

**Pull request merged**
